### PR TITLE
Make tedge a multi-call binary

### PR DESCRIPTION
**Pull request by jarhodes314**
## Proposed changes
This is the first change related to #2281. This converts tedge into a multicall (aka busybox-style) binary, so `tedge` will behave differently depending on the executable name (e.g. creating a symlink to `tedge` called `tedge-mapper` and then executing tedge via this symlink will result in the `tedge-mapper` cli being called, not `tedge`).

The PR includes:
* Creation of symlinks for each of the tedge binaries via the `tedge init` command (e.g.  `/usr/bin/tedge-mapper` is a symlink to `/usr/bin/tedge`) 
* Updated packaging of tar and linux packages to utilize the multi-call `tedge` binary instead of including a binary in each individual package. Now the linux packages only contains some configuration files and service definitions (and some unrelated backwards compatible symlinks)

**NB** the changes will be clearer if you view the diff of the two commits separately as I renamed a bunch of main.rs files to lib.rs in the first commit, and then added a bunch of new main.rs files to preserve backwards compatibility. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/2281

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by didier-wenzek** [comment]
This is a nice disk foot-print improvement!

With this PR, `tedge` wraps in a single executable the `tedge` command with `tedge-agent`, `tedge-mapper` and some of the plugins. 

Before:

```
 ls -lh target/release/tedge*
-rwxrwxr-x 2 didier didier 6,4M sept. 28 20:53 target/release/tedge
-rwxrwxr-x 2 didier didier 6,7M sept. 28 20:53 target/release/tedge-agent
-rwxrwxr-x 2 didier didier 3,0M sept. 28 20:53 target/release/tedge-apt-plugin
-rwxrwxr-x 2 didier didier 6,6M sept. 28 20:53 target/release/tedge-configuration-plugin
-rwxrwxr-x 2 didier didier 6,1M sept. 28 20:53 target/release/tedge-log-plugin
-rwxrwxr-x 2 didier didier 7,9M sept. 28 20:53 target/release/tedge-mapper
-rwxrwxr-x 2 didier didier 4,7M sept. 28 20:53 target/release/tedge-watchdog
-rwxrwxr-x 2 didier didier 6,4M sept. 28 20:53 target/release/c8y-firmware-plugin
-rwxrwxr-x 2 didier didier 3,5M sept. 28 20:53 target/release/c8y-remote-access-plugin
```

After:

```
$ ls -lh target/release/tedge
-rwxrwxr-x 2 didier didier 9,7M sept. 28 21:02 target/release/tedge

$ ln -s $PWD/target/release/tedge /tmp/tedge-mapper
$ /tmp/tedge-mapper
tedge-mapper is the mapper that translates thin-edge.io data model to c8y/az data model.

Usage: tedge-mapper [OPTIONS] <COMMAND>

Commands:
  az        
  aws       
  c8y       
  collectd  
  help      Print this message or the help of the given subcommand(s)
```

__Note__ this PR still provides old equivalent executables for `tedge-agent`, `tedge-mapper`, plugins ... This is only a temporary step before packaging is updated.

That said, I see this as first step. There are still too many commands and daemons for the users to install and manage.
One need to go a step further and include in `tedge-agent` all the `tedge-xxx-plugin`s: for log management, as well as for configuration and firmware management (not implemented yet). This can simply be done by integrating the associated actors with those of the `tedge-agent`.

**Review by didier-wenzek — COMMENTED** [review]
I'm okay to go in that direction and even to merge this PR.

However:
- we must wait an approval from @reubenmiller as this has an important impact on packaging.
- we must update packaging and deprecate the executables which are currently provided for backward compatibilities.
- We still have to extend `tedge-agent` to include all the device management features and to deprecate the `tedge-xxx-plugin`s.

**Comment by reubenmiller** [comment]
> I'm okay to go in that direction and even to merge this PR.
> 
> However:
> 
> * we must wait an approval from @reubenmiller as this has an important impact on packaging.
> * we must update packaging and deprecate the executables which are currently provided for backward compatibilities.
> * We still have to extend `tedge-agent` to include all the device management features and to deprecate the `tedge-xxx-plugin`s.

I had a quick look and it's going in a great direction. I'm happy to merge this even though it will lead to a brief period (about a week) where the total binary size would be ~2.5MB more than before until we switch to the multi call binary.

Additional follow up steps might be able to assist the follow up packaging aspects (in addition to the points that @didier-wenzek already mentioned):

* extend the `tedge init` command to create the symlinks for the required binaries (and also take care of upgrade paths when we consolidate the tedge-xxx-plugin binaries as Didier mentioned)
* Add a new subcommand to remove the initialization (but we can post pone this work until a later point in time)

**Comment by reubenmiller** [comment]
@jarhodes314 This is looking really good now. Can you also add support for `tedge-apt-plugin`? I know it was originally out of scope, however this is the only binary which is not included, and it would simplify the packaging slightly and decrease the overall overhead on Debian systems.

Afterwards, I'll push the packaging changes required to make use of the multi-call binary so that we can have the packaging sorted before merging this PR (and validation from the system tests that everything works as before).

**Comment by jarhodes314** [comment]
> @jarhodes314 This is looking really good now. Can you also add support for `tedge-apt-plugin`? I know it was originally out of scope, however this is the only binary which is not included, and it would simplify the packaging slightly and decrease the overall overhead on Debian systems.
> 
> Afterwards, I'll push the packaging changes required to make use of the multi-call binary so that we can have the packaging sorted before merging this PR (and validation from the system tests that everything works as before).

tedge-apt-plugin is already supported, but it's dealt with directly in main.rs so the parsing can be done separately (as the plugin does some error handling for itself so it can change the default exit codes).

**Comment by jarhodes314** [comment]
> @jarhodes314 This is looking really good now. Can you also add support for `tedge-apt-plugin`? I know it was originally out of scope, however this is the only binary which is not included, and it would simplify the packaging slightly and decrease the overall overhead on Debian systems.
> 
> Afterwards, I'll push the packaging changes required to make use of the multi-call binary so that we can have the packaging sorted before merging this PR (and validation from the system tests that everything works as before).

tedge-apt-plugin is already supported, but it's dealt with directly in main.rs so the parsing can be done separately (as the plugin does some error handling for itself so it can change the default exit codes).

**Comment by reubenmiller** [comment]
> > @jarhodes314 This is looking really good now. Can you also add support for `tedge-apt-plugin`? I know it was originally out of scope, however this is the only binary which is not included, and it would simplify the packaging slightly and decrease the overall overhead on Debian systems.
> > Afterwards, I'll push the packaging changes required to make use of the multi-call binary so that we can have the packaging sorted before merging this PR (and validation from the system tests that everything works as before).
> 
> tedge-apt-plugin is already supported, but it's dealt with directly in main.rs so the parsing can be done separately (as the plugin does some error handling for itself so it can change the default exit codes).

Though I don't think the `tedge init` command created the symlink to `/usr/bin/tedge-apt-plugin` then...

**Comment by reubenmiller** [comment]
> > > @jarhodes314 This is looking really good now. Can you also add support for `tedge-apt-plugin`? I know it was originally out of scope, however this is the only binary which is not included, and it would simplify the packaging slightly and decrease the overall overhead on Debian systems.
> > > Afterwards, I'll push the packaging changes required to make use of the multi-call binary so that we can have the packaging sorted before merging this PR (and validation from the system tests that everything works as before).
> > 
> > 
> > tedge-apt-plugin is already supported, but it's dealt with directly in main.rs so the parsing can be done separately (as the plugin does some error handling for itself so it can change the default exit codes).
> 
> Though I don't think the `tedge init` command created the symlink to `/usr/bin/tedge-apt-plugin` then...

Actually `/usr/bin/tedge-apt-plugin` does exist...something strange is happening...I'll recheck it now.

File: crates/core/tedge/src/main.rs [hunk 1]
```diff
@@ -3,37 +3,86 @@
...
 fn main() -> anyhow::Result<()> {
-    set_log_level(tracing::Level::WARN);
+    let executable_name = executable_name();
 
-    let opt = tedge::cli::Opt::parse();
-
-    if opt.init {
-        warn!("This --init option has been deprecated and will be removed in a future release. Use the `tedge init` command instead");
-        return Ok(());
+    if executable_name.as_deref() == Some("apt") {
```
Comment (reubenmiller) [inline crates/core/tedge/src/main.rs:18]: Can you change it to also look for the full name of the binary, `tedge-apt-plugin` as well?

For example (obviously modify the code, but you get the idea):

```
if executable_name.as_deref() == Some("apt") || executable_name.as_deref() == Some("tedge-apt-plugin") {
```

Then for the symlink creation, I would create the symlinks under:
* `/usr/bin/tedge` -> `/usr/bin/tedge-apt-plugin` (we can't use /usr/bin/apt as this is already reserved by the real apt binary)

Then the actual tedge-apt-plugin package can handle the creation of the `/etc/tedge/sm-plugins/apt` symlink when the package is installed.

Comment (jarhodes314) [reply crates/core/tedge/src/main.rs:18]: Ah sure, I overlooked that

Comment (reubenmiller) [reply crates/core/tedge/src/main.rs]: Looks good now.

**Review by reubenmiller — APPROVED** [review]
Great step forward. Package sizes are reduced by approximately 4 times!

Approved.

File: crates/core/tedge/src/cli/init.rs [hunk 2]
```diff
@@ -1,26 +1,102 @@
...
+                        Path::new("./tedge")
+                    } else {
+                        &*executable_name
+                    };
+                    std::os::unix::fs::symlink(tedge, &link).with_context(|| {
+                        format!("creating symlink for {component} to {}", tedge.display())
+                    })?;
+
+                    let res = std::process::Command::new("chown")
+                        .arg("-h")
```
Comment (Bravo555) [inline crates/core/tedge/src/cli/init.rs:80]: nit: Surprised that in chown, `-h` is not `--help`. For readability I think it's be better to use long flags, so `--no-dereference` in this case.

File: crates/core/tedge/src/cli/init.rs [hunk 3]
```diff
@@ -1,26 +1,102 @@
...
+                    if file_exists {
+                        nix::unistd::unlink(&link).with_context(|| {
+                            format!("removing old version of {component} at {}", link.display())
+                        })?;
+                    }
+
+                    let tedge = if self.relative_links {
+                        // TODO this could lead to broken links
+                        // TODO is this feature even that useful?
+                        Path::new("./tedge")
```
Comment (Bravo555) [inline crates/core/tedge/src/cli/init.rs:71]: question: What are some use cases where using relative links could be useful?

**Review by Bravo555 — APPROVED** [review]
One question, but otherwise no issues. This should also simplify Yocto Project recipes quite a lot. Nice work!

File: crates/core/tedge/src/cli/init.rs (hunk 3 above)
Comment (jarhodes314) [reply crates/core/tedge/src/cli/init.rs:71]: Honestly, I'm not entirely sure, hence the comment immediately above it. I guess it makes moving the binaries a tad easier, but the fact that `tedge init` sets the links up should solve that problem in any case. I guess if someone wants to store tedge on some non-deterministically named mount point or something that might be a possible use case, but I am leaning towards declaring this feature pointless and removing it to simplify the code and CLI.

Comment (jarhodes314) [reply crates/core/tedge/src/cli/init.rs:71]: I've slightly refactored this, but I think I'll keep this, as it doesn't add much complexity, and there may be valid use cases. Absolute paths are probably the right default, and what `tedge-apt-plugin` uses (although for that case, a relative path wouldn't make any sense).

**Pull request merged**